src/rgw/rgw_checksum.h

```diff
@@ -0,0 +1,103 @@
+#ifndef CEPH_RGW_CHECKSUM_H
+#define CEPH_RGW_CHECKSUM_H
+#include "common/armor.h"
+#include "common/ceph_context.h"
+#include "common/ceph_crypto.h"
+#include <boost/crc.hpp>
+#include <string>
+#include <map>
+using namespace std;
+
+#define RGW_CHECKSUM_CRC32_DIGESTSIZE 4
+#define RGW_CHECKSUM_CRC32_ARMORED_SIZE 8
+#define RGW_CHECKSUM_SHA1_DIGESTSIZE 20
+#define RGW_CHECKSUM_SHA1_ARMORED_SIZE 28
+
+typedef boost::crc_optimal<32, 0x04C11DB7, 0xFFFFFFFF, 0xFFFFFFFF, true, true>
+    crc32_type;
+
+class RGWChecksum {
+  bool need_calc_crc32;
+  char supplied_crc32_bin[RGW_CHECKSUM_CRC32_DIGESTSIZE + 1];
+  char final_crc32_bin[RGW_CHECKSUM_CRC32_DIGESTSIZE + 1];
+  char final_crc32_str[RGW_CHECKSUM_CRC32_DIGESTSIZE * 2 + 1];
+  char resp_crc32_bin[RGW_CHECKSUM_CRC32_DIGESTSIZE * 2 + 16];
+  char resp_crc32_b64[RGW_CHECKSUM_CRC32_DIGESTSIZE * 2 + 16];
+  crc32_type hash_crc32;
+  bool need_calc_sha1;
+  char supplied_sha1_bin[RGW_CHECKSUM_SHA1_DIGESTSIZE + 1];
+  char final_sha1_bin[RGW_CHECKSUM_SHA1_DIGESTSIZE + 1];
+  char final_sha1_str[RGW_CHECKSUM_SHA1_DIGESTSIZE * 2 + 1];
+  char resp_sha1_bin[RGW_CHECKSUM_SHA1_DIGESTSIZE * 2 + 16];
+  char resp_sha1_b64[RGW_CHECKSUM_SHA1_DIGESTSIZE * 2 + 16];
+  ceph::crypto::SHA1 hash_sha1;
```

Comment:
as a point of design feedback, we should consider some kind of variant-like abstraction here. we'll only ever have one algorithm enabled, so we shouldn't need to have all of their buffers in memory. this will become more of an issue as we add more algorithms

Reply:
I think https://github.com/ceph/ceph/pull/30606/commits/c56fc3be1572c8ec0d703b3a8b691d767f119add did a good job of packaging things up compactly and minimizing overhead, using only modern apis